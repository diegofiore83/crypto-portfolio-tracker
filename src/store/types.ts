export interface CryptoHolding {
  name: string;
  symbol: string;
  quantity: number;
  currentPrice: number;
  totalValue: number;
}
