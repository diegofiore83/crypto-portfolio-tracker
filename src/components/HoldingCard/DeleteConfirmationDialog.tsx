import React from "react";
import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
} from "@mui/material";

interface DeleteConfirmationDialogProps {
  open: boolean;
  onCancel: () => void;
  onConfirm: () => void;
  name: string;
  symbol: string;
}

const DeleteConfirmationDialog: React.FC<DeleteConfirmationDialogProps> = ({
  open,
  onCancel,
  onConfirm,
  name,
  symbol,
}) => (
  <Dialog open={open} onClose={onCancel} aria-labelledby="alert-dialog-title">
    <DialogTitle id="alert-dialog-title">
      Delete {name} ({symbol.toUpperCase()})
    </DialogTitle>
    <DialogContent>
      <DialogContentText>
        Are you sure you want to delete this holding? This action cannot be
        undone.
      </DialogContentText>
    </DialogContent>
    <DialogActions>
      <Button onClick={onCancel} color="primary">
        Cancel
      </Button>
      <Button onClick={onConfirm} color="error" autoFocus>
        Delete
      </Button>
    </DialogActions>
  </Dialog>
);

export default DeleteConfirmationDialog;
