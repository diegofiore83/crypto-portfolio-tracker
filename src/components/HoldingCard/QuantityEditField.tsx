import React from "react";
import { Box, Button, TextField } from "@mui/material";

interface QuantityEditFieldProps {
  editedQuantity: string;
  error: string | null;
  handleQuantityChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  handleSubmit: () => void;
  isValidQuantity: boolean;
}

const QuantityEditField: React.FC<QuantityEditFieldProps> = ({
  editedQuantity,
  error,
  handleQuantityChange,
  handleSubmit,
  isValidQuantity,
}) => (
  <Box display="flex" gap={2} alignItems="flex-start">
    <TextField
      label="Quantity"
      size="small"
      type="number"
      value={editedQuantity}
      onChange={handleQuantityChange}
      onKeyDown={(e) => e.key === "Enter" && isValidQuantity && handleSubmit()}
      error={!!error}
      helperText={error}
      sx={{ flexGrow: 1 }}
    />
    <Button
      color="primary"
      variant="contained"
      onClick={handleSubmit}
      disabled={!isValidQuantity}
    >
      Save
    </Button>
  </Box>
);

export default QuantityEditField;
