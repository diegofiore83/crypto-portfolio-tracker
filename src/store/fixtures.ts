import { CryptoAsset, MarketData } from "./types";

export const mockCryptoAssets: CryptoAsset[] = [
  {
    id: "bitcoin",
    symbol: "btc",
    name: "Bitcoin",
    image: "https://example.com/bitcoin.png",
    current_price: 45000,
    market_cap: 850000000000,
    market_cap_rank: 1,
    total_volume: 35000000000,
    high_24h: 46000,
    low_24h: 44000,
    price_change_24h: -500,
    price_change_percentage_24h: -1.1,
    market_cap_change_24h: -10000000000,
    market_cap_change_percentage_24h: -1.2,
    circulating_supply: 19000000,
    total_supply: 21000000,
    max_supply: 21000000,
    ath: 65000,
    ath_change_percentage: -30,
    ath_date: "2021-04-14T00:00:00.000Z",
    atl: 67.81,
    atl_change_percentage: 66300,
    atl_date: "2013-07-06T00:00:00.000Z",
    roi: null,
    last_updated: "2021-09-01T00:00:00.000Z",
    fully_diluted_valuation: 850000000000,
  },
  {
    id: "ethereum",
    symbol: "eth",
    name: "Ethereum",
    image: "https://example.com/ethereum.png",
    current_price: 3000,
    market_cap: 350000000000,
    market_cap_rank: 2,
    total_volume: 20000000000,
    high_24h: 3100,
    low_24h: 2900,
    price_change_24h: -50,
    price_change_percentage_24h: -1.6,
    market_cap_change_24h: -5000000000,
    market_cap_change_percentage_24h: -1.4,
    circulating_supply: 115000000,
    total_supply: 115000000,
    max_supply: null,
    ath: 4300,
    ath_change_percentage: -30,
    ath_date: "2021-05-12T00:00:00.000Z",
    atl: 0.432979,
    atl_change_percentage: 692000,
    atl_date: "2015-10-20T00:00:00.000Z",
    roi: null,
    last_updated: "2021-09-01T00:00:00.000Z",
    fully_diluted_valuation: 350000000000,
  },
];

export const mockMarketData: MarketData = {
  prices: [
    [1738530045736, 97793.53821562871],
    [1738533735812, 96988.74415938454],
    [1738537427863, 97815.84535257479],
  ],
  market_caps: [
    [1738530045736, 1937072986217.3457],
    [1738533735812, 1921781912853.0679],
    [1738537427863, 1936077131894.8628],
  ],
  total_volumes: [
    [1738530045736, 0],
    [1738533735812, 0],
    [1738537427863, 0],
  ],
};
