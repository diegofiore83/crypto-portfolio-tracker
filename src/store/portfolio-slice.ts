import { createSlice, PayloadAction } from "@reduxjs/toolkit";
import { CryptoHolding } from "./types";

const loadFromLocalStorage = (): CryptoHolding[] => {
  const data = localStorage.getItem("portfolio");
  return data ? JSON.parse(data) : [];
};

const saveToLocalStorage = (state: CryptoHolding[]) => {
  localStorage.setItem("portfolio", JSON.stringify(state));
};

const initialState: CryptoHolding[] = loadFromLocalStorage();

const portfolioSlice = createSlice({
  name: "portfolio",
  initialState,
  reducers: {
    addCrypto: (state, action: PayloadAction<CryptoHolding>) => {
      state.push(action.payload);
      saveToLocalStorage(state);
    },
    editCrypto: (state, action: PayloadAction<CryptoHolding>) => {
      const index = state.findIndex(
        (crypto) => crypto.symbol === action.payload.symbol
      );
      if (index !== -1) {
        state[index] = action.payload;
        saveToLocalStorage(state);
      }
    },
    deleteCrypto: (state, action: PayloadAction<string>) => {
      const newState = state.filter(
        (crypto) => crypto.symbol !== action.payload
      );
      saveToLocalStorage(newState);
      return newState;
    },
  },
});

export const { addCrypto, editCrypto, deleteCrypto } = portfolioSlice.actions;
export default portfolioSlice.reducer;
