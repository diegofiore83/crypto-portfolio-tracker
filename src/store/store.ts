import { configureStore } from "@reduxjs/toolkit";
import cryptoReducer from "./slices/crypto-slice";
import portfolioReducer from "./slices/portfolio-slice";

const store = configureStore({
  reducer: {
    crypto: cryptoReducer,
    portfolio: portfolioReducer,
  },
});

// TypeScript types for Redux
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;

export default store;
