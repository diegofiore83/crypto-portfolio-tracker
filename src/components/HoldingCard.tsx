import React, { useCallback, useState } from "react";
import {
  Box,
  Divider,
  Grid2 as Grid,
  Card,
  CardActions,
  CardContent,
  CardHeader,
  Avatar,
  Button,
  Typography,
} from "@mui/material";
import CryptoQuantityField from "./HoldingCard/QuantityEditField";
import DeleteConfirmationDialog from "./HoldingCard/DeleteConfirmationDialog";
import { useAppDispatch } from "../store/hooks";
import { deleteCrypto, editCrypto } from "../store/slices/portfolio-slice";

interface HoldingCardProps {
  id: string;
  name: string;
  symbol: string;
  image: string;
  quantity: number;
  current_price: number;
  total_value: number;
}

const HoldingCard: React.FC<HoldingCardProps> = ({
  id,
  name,
  symbol,
  image,
  quantity,
  current_price,
  total_value,
}) => {
  const [isEditingMode, setIsEditingMode] = useState<boolean>(false);
  const [editedQuantity, setEditedQuantity] = useState<string>(
    quantity.toString()
  );
  const [error, setError] = useState<string | null>(null);
  const [showDeleteConfirmation, setShowDeleteConfirmation] =
    useState<boolean>(false);
  const dispatch = useAppDispatch();

  const isValidQuantity =
    editedQuantity !== "" && parseFloat(editedQuantity) > 0;

  const handleDeleteClick = () => {
    setShowDeleteConfirmation(true);
  };

  const handleDeleteCancel = () => {
    setShowDeleteConfirmation(false);
  };

  const handleDeleteConfirm = useCallback(() => {
    dispatch(deleteCrypto(id));
    setShowDeleteConfirmation(false);
  }, [dispatch, id]);

  const handleQuantityChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value.trim();
    const numValue = parseFloat(value);

    setEditedQuantity(value);
    setError(
      value === ""
        ? null
        : isNaN(numValue) || numValue <= 0
        ? "Quantity must be positive"
        : null
    );
  };

  const handleSubmit = useCallback(() => {
    if (isValidQuantity) {
      dispatch(editCrypto({ id, quantity: parseFloat(editedQuantity) }));
      setEditedQuantity(editedQuantity);
      setIsEditingMode(false);
    }
  }, [dispatch, id, editedQuantity, isValidQuantity]);

  const handleCancel = useCallback(() => {
    setIsEditingMode(false);
    setEditedQuantity(quantity.toString());
    setError(null);
  }, [quantity]);

  return (
    <Grid size={{ xs: 12, sm: 6, md: 4, lg: 3 }}>
      <Card>
        <CardHeader
          avatar={
            <Avatar
              alt={name}
              src={image}
              sx={{ width: 32, height: 32, mb: 1 }}
            />
          }
          title={
            <Box
              display="flex"
              alignItems="center"
              justifyContent="space-between"
            >
              <Typography fontWeight="bold">{symbol.toUpperCase()}</Typography>
            </Box>
          }
          subheader={
            <Typography variant="subtitle2" noWrap maxWidth="90%">
              {name}
            </Typography>
          }
        />
        <CardContent>
          <Box minHeight={85}>
            {isEditingMode ? (
              <CryptoQuantityField
                editedQuantity={editedQuantity}
                error={error}
                handleQuantityChange={handleQuantityChange}
                handleSubmit={handleSubmit}
                isValidQuantity={isValidQuantity}
              />
            ) : (
              <>
                <Box
                  display="flex"
                  justifyContent="space-between"
                  alignItems="center"
                  mb={0.5}
                >
                  <Typography variant="body2" color="text.secondary">
                    Quantity
                  </Typography>
                  <Typography variant="body1" fontWeight="medium">
                    {quantity.toLocaleString(undefined, {
                      minimumFractionDigits: 0,
                      maximumFractionDigits: 5,
                      useGrouping: true,
                    })}
                  </Typography>
                </Box>

                <Box
                  display="flex"
                  justifyContent="space-between"
                  alignItems="center"
                  mb={0.5}
                >
                  <Typography variant="body2" color="text.secondary">
                    Current Price
                  </Typography>
                  <Typography variant="body1" fontWeight="medium">
                    $
                    {current_price.toLocaleString(undefined, {
                      minimumFractionDigits: 0,
                      maximumFractionDigits: 5,
                      useGrouping: true,
                    })}
                  </Typography>
                </Box>

                <Divider sx={{ my: 1 }} />

                <Box
                  display="flex"
                  justifyContent="space-between"
                  alignItems="center"
                >
                  <Typography
                    variant="body1"
                    fontWeight="medium"
                    color="text.primary"
                  >
                    Total Value
                  </Typography>
                  <Typography
                    variant="body1"
                    fontWeight="bold"
                    color="primary.main"
                  >
                    $
                    {total_value.toLocaleString(undefined, {
                      minimumFractionDigits: 2,
                      maximumFractionDigits: 2,
                    })}
                  </Typography>
                </Box>
              </>
            )}
          </Box>
        </CardContent>
        <CardActions>
          <Box
            display="flex"
            justifyContent="space-between"
            alignItems="center"
            width="100%"
            paddingRight={2}
          >
            <Button
              size="small"
              color="primary"
              onClick={() =>
                isEditingMode ? handleCancel() : setIsEditingMode(true)
              }
            >
              {isEditingMode ? "Cancel" : `Edit ${symbol.toUpperCase()}`}
            </Button>
            <Button size="small" color="error" onClick={handleDeleteClick}>
              Delete {symbol.toUpperCase()}
            </Button>
          </Box>
        </CardActions>
      </Card>

      <DeleteConfirmationDialog
        open={showDeleteConfirmation}
        onCancel={handleDeleteCancel}
        onConfirm={handleDeleteConfirm}
        name={name}
        symbol={symbol}
      />
    </Grid>
  );
};

export default HoldingCard;
