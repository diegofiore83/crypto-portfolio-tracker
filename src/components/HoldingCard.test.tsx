import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import { configureStore } from "@reduxjs/toolkit";
import { MemoryRouter } from "react-router-dom";
import { Provider } from "react-redux";
import portfolioReducer from "../store/slices/portfolio-slice";
import { CryptoHolding } from "../store/types";
import HoldingCard from "./HoldingCard";
import { RootState } from "../store/store";

const mockProps = {
  id: "bitcoin",
  name: "Bitcoin",
  symbol: "BTC",
  image: "https://example.com/bitcoin.png",
  quantity: 2,
  current_price: 50000,
  total_value: 100000,
};

describe("HoldingCard Component", () => {
  let store: ReturnType<typeof configureStore>;

  beforeEach(() => {
    store = configureStore({
      reducer: { portfolio: portfolioReducer },
      preloadedState: {
        portfolio: [{ id: "bitcoin", quantity: 2 } as CryptoHolding],
      },
    });
  });

  const renderWithProvider = (component: React.ReactElement) =>
    render(
      <Provider store={store}>
        <MemoryRouter>{component}</MemoryRouter>
      </Provider>
    );

  test("renders HoldingCard correctly (snapshot)", () => {
    const { asFragment } = renderWithProvider(<HoldingCard {...mockProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  test("displays correct initial data", () => {
    renderWithProvider(<HoldingCard {...mockProps} />);
    expect(screen.getByText("Bitcoin")).toBeInTheDocument();
    expect(screen.getByText("BTC")).toBeInTheDocument();
    expect(screen.getByText(/\$50,000(\.00)?/)).toBeInTheDocument();
    expect(screen.getByText(/\$100,000(\.00)?/)).toBeInTheDocument();
  });

  test("opens edit mode and updates quantity", async () => {
    renderWithProvider(<HoldingCard {...mockProps} />);

    fireEvent.click(screen.getByText("Edit BTC"));
    fireEvent.change(screen.getByLabelText("Quantity"), {
      target: { value: "3" },
    });
    fireEvent.click(screen.getByText("Save"));

    await waitFor(() => {
      const updatedState = (store.getState() as RootState).portfolio;
      expect(updatedState).toContainEqual({ id: "bitcoin", quantity: 3 });
    });
  });

  test("shows delete confirmation dialog", async () => {
    renderWithProvider(<HoldingCard {...mockProps} />);
    fireEvent.click(screen.getByText("Delete BTC"));

    await waitFor(() => {
      expect(
        screen.getByRole("dialog", { name: /delete bitcoin/i })
      ).toBeInTheDocument();
    });
  });

  test("deletes the holding when confirmed", async () => {
    renderWithProvider(<HoldingCard {...mockProps} />);

    fireEvent.click(screen.getByText("Delete BTC"));
    fireEvent.click(screen.getByText("Delete"));

    await waitFor(() => {
      const updatedState = (store.getState() as RootState).portfolio;
      expect(updatedState).not.toContainEqual({ id: "bitcoin", quantity: 2 });
    });
  });
});
