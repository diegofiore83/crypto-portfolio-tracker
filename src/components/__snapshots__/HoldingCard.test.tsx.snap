// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HoldingCard Component renders HoldingCard correctly (snapshot) 1`] = `
<DocumentFragment>
  <div
    class="MuiGrid2-root MuiGrid2-direction-xs-row MuiGrid2-grid-xs-12 MuiGrid2-grid-sm-6 MuiGrid2-grid-md-4 MuiGrid2-grid-lg-3 css-bz5kom-MuiGrid2-root"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1j74uew-MuiPaper-root-MuiCard-root"
      style="--Paper-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);"
    >
      <div
        class="MuiCardHeader-root css-149tc0y-MuiCardHeader-root"
      >
        <div
          class="MuiCardHeader-avatar css-1r9wl67-MuiCardHeader-avatar"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular css-ci7oj7-MuiAvatar-root"
          >
            <img
              alt="Bitcoin"
              class="MuiAvatar-img css-1su80sj-MuiAvatar-img"
              src="https://example.com/bitcoin.png"
            />
          </div>
        </div>
        <div
          class="MuiCardHeader-content css-1jkp9bu-MuiCardHeader-content"
        >
          <span
            class="MuiTypography-root MuiTypography-body2 MuiCardHeader-title css-7ctihp-MuiTypography-root"
          >
            <div
              class="MuiBox-root css-1lekzkb"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-lhsbas-MuiTypography-root"
              >
                BTC
              </p>
            </div>
          </span>
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-noWrap css-14ggmo7-MuiTypography-root"
          >
            Bitcoin
          </h6>
        </div>
      </div>
      <div
        class="MuiCardContent-root css-1lt5qva-MuiCardContent-root"
      >
        <div
          class="MuiBox-root css-1y359bf"
        >
          <div
            class="MuiBox-root css-n64gf7"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
            >
              Quantity
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-rt9437-MuiTypography-root"
            >
              2
            </p>
          </div>
          <div
            class="MuiBox-root css-n64gf7"
          >
            <p
              class="MuiTypography-root MuiTypography-body2 css-1wle3ir-MuiTypography-root"
            >
              Current Price
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-rt9437-MuiTypography-root"
            >
              $50,000
            </p>
          </div>
          <hr
            class="MuiDivider-root MuiDivider-fullWidth css-hv6zls-MuiDivider-root"
          />
          <div
            class="MuiBox-root css-69i1ev"
          >
            <p
              class="MuiTypography-root MuiTypography-body1 css-65t5jw-MuiTypography-root"
            >
              Total Value
            </p>
            <p
              class="MuiTypography-root MuiTypography-body1 css-sk0gwb-MuiTypography-root"
            >
              $100,000.00
            </p>
          </div>
        </div>
      </div>
      <div
        class="MuiCardActions-root MuiCardActions-spacing css-1q4nm6f-MuiCardActions-root"
      >
        <div
          class="MuiBox-root css-1kms82h"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-1ccsd5i-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Edit BTC
          </button>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorError MuiButton-root MuiButton-text MuiButton-textError MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorError css-1nrx54a-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Delete BTC
          </button>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
